### PR TITLE
🐛 prevent hang when HelloRetryRequest is returned

### PR DESCRIPTION
**Pull request by jaym**
It was possible for the following query to hang
```
tls.ciphers
```

This happened in cases where the server response with a `HelloRetryRequest`. This looks like a `ServerHello` until you check the random field of that message.

We ended up hanging because we thought the server replied successfully with a hello and were waiting for more data, and the server was waiting for us to write.

### CODE REVIEW

File: providers/network/resources/tlsshake/tlsshake.go [hunk 1]
```diff
@@ -309,12 +309,24 @@ func (s *Tester) parseAlert(data []byte, conf *ScanConfig) error {
 	return nil
 }
 
+var helloRetryRequestRandom = []byte{
```
Comment (czunker) [inline providers/network/resources/tlsshake/tlsshake.go:318]: nit: Perhaps add a comment or rename it.

Only the comment in the PR helped me understand that this is not random data, but the random message in the response.

File: providers/network/resources/tlsshake/tlsshake.go [hunk 2]
```diff
@@ -309,12 +309,24 @@ func (s *Tester) parseAlert(data []byte, conf *ScanConfig) error {
...
 
-	idx += 2 + 32
-	// handshake tls version (2), which we don't need yet (we will look at it in the extension if necessary)
-	// random (32), which we don't need
+	idx += 2
+	random := data[idx : idx+32]
+	if bytes.Equal(random, helloRetryRequestRandom) {
+		// The server wants us to retry. However, we hardcode a bunch of inputs
+		// so we can't really retry.
+		return errors.New("unexpected HelloRetryRequest")
```
Comment (czunker) [inline providers/network/resources/tlsshake/tlsshake.go]: Is it possible to extract the key_share or supported_versions from the response extensions?

That might help the next we hit this, and we need to figure out what else besides `secp256r1` is needed.

**Review by czunker — APPROVED** [review]

File: providers/network/resources/tlsshake/tlsshake.go (hunk 2 above)
Comment (jaym) [reply providers/network/resources/tlsshake/tlsshake.go]: I've added the key share group to the error message. We already seem to do something with supported versions

File: providers/network/resources/tlsshake/tlsshake.go (hunk 1 above)
Comment (jaym) [reply providers/network/resources/tlsshake/tlsshake.go:318]: Added a comment

**Review by czunker — APPROVED** [review]
Really nice! Thanks @jaym

**Pull request merged**